File: leetcode/algorithms/convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.cpp
#include <vector>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *sortedArrayToBST(vector<int> &nums) {
    return buildBST(nums, 0, nums.size() - 1);
  }
  TreeNode *buildBST(vector<int> &nums, int i, int j) {
    if (i > j) {
      return nullptr;
    }
    if (i == j) {
      return new TreeNode(nums[i]);
    }
    size_t mid = (j - i) / 2 + i;
    TreeNode *node = new TreeNode(nums[mid]);
    node->left = buildBST(nums, i, mid - 1);
    node->right = buildBST(nums, mid + 1, j);
    return node;
  }
};

File: leetcode/algorithms/convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def helper(i: int, j: int):
            if i > j:
                return None

            if i == j:
                return TreeNode(val=nums[i])

            mid = (j - i) // 2 + i
            mid_node = TreeNode(val=nums[mid])

            mid_node.left = helper(i, mid - 1)
            mid_node.right = helper(mid + 1, j)

            return mid_node

        return helper(0, len(nums) - 1)
